src/main/java/net/neoforged/neoforge/common/util/GlobalVec3.java

```diff
@@ -5,51 +5,295 @@
+        this(dimension, new WeakReference<>(null), x, y, z);
+    }
+
     public GlobalVec3(Level level, Vector3f vector3f) {
-        super(vector3f);
-        this.level = level;
+        this(level.dimension(), new WeakReference<>(level), vector3f.x, vector3f.y, vector3f.z);
+    }
+
+    public GlobalVec3(ResourceKey<Level> dimension, Vector3f vector3f) {
+        this(dimension, new WeakReference<>(null), vector3f.x, vector3f.y, vector3f.z);
+    }
+
+    protected GlobalVec3(ResourceKey<Level> dimension, WeakReference<Level> level, Vec3 vec3) {
+        this(dimension, level, vec3.x, vec3.y, vec3.z);
     }
 
     public GlobalVec3(Level level, Vec3 vec3) {
-        super(vec3.x, vec3.y, vec3.z);
-        this.level = level;
+        this(level.dimension(), new WeakReference<>(level), vec3);
+    }
+
+    public GlobalVec3(ResourceKey<Level> dimension, Vec3 vec3) {
+        this(dimension, new WeakReference<>(null), vec3);
     }
 
     public GlobalVec3(Entity entity) {
         this(entity.level(), entity.position());
     }
 
     public GlobalVec3(Level level, BlockPos blockpos) {
-        this(level, blockpos.getX(), blockpos.getY(), blockpos.getZ());
+        this(level.dimension(), new WeakReference<>(level), blockpos.getX(), blockpos.getY(), blockpos.getZ());
+    }
+
+    public GlobalVec3(ResourceKey<Level> dimension, BlockPos blockpos) {
+        this(dimension, new WeakReference<>(null), blockpos.getX(), blockpos.getY(), blockpos.getZ());
+    }
+
+    public GlobalVec3(GlobalPos globalpos) {
+        this(globalpos.dimension(), globalpos.pos());
     }
 
+    // setters
+
     public GlobalVec3 withX(double x) {
-        return new GlobalVec3(level, x, y, z);
+        return new GlobalVec3(dimension, level, x, y, z);
     }
 
     public GlobalVec3 withY(double y) {
-        return new GlobalVec3(level, x, y, z);
+        return new GlobalVec3(dimension, level, x, y, z);
     }
 
     public GlobalVec3 withZ(double z) {
-        return new GlobalVec3(level, x, y, z);
+        return new GlobalVec3(dimension, level, x, y, z);
     }
 
     public GlobalVec3 withLevel(Level level) {
-        return new GlobalVec3(level, x, y, z);
+        return new GlobalVec3(level.dimension(), new WeakReference<>(level), x, y, z);
+    }
+
+    public GlobalVec3 withDimension(ResourceKey<Level> dimension) {
+        return new GlobalVec3(dimension, new WeakReference<>(null), x, y, z);
+    }
+
+    // getters
+
+    public ResourceKey<Level> getDimension() {
+        return dimension;
+    }
+
+    /**
+     * Returns the level of this global position if it is available.
+     * <p>
+     * Use {@link #getLevel(MinecraftServer)} or {@link #getLevel(Level)} if possible.
+     * 
+     * @return The level.
+     */
+    public @Nullable Level getLevel() {
+        return level.get();
+    }
+
+    public Level getLevel(Level anotherLevel) {
+        return getLevel(anotherLevel.getServer());
```

Comment:
`Level#getServer()` may return null, but it is being passed to `#getLevel(MinecraftServer)` which expects a non-null server. I think we should just remove this getter and reinforce that `getLevel(MinecraftServer)` requires a non-null server instance, so it is on the user to determine if the server instance is available to them.

Reply:
I modelled this after the uses of GlobalPos, all of which run this unconditionally...oh, but I made the parameter a Level, not a ServerLevel. Oops, that's the issue.

Fixed.